File: analyze.ts
/**
 * Calcutta Analysis Tool — Youche CC
 *
 * Pulls last 20 rounds for each player from GHIN, computes:
 *   - Scoring differential trend vs posted handicap index
 *   - Sandbagger score (how often they beat their handicap significantly)
 *   - Volatility (standard deviation of differentials)
 *   - Flight assignment (A / B / C)
 *
 * Usage:
 *   cp .env.example .env && fill in creds
 *   npm install
 *   npm run analyze
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { GhinClient } from 'ghin';
import Papa from 'papaparse';

// ── Config ────────────────────────────────────────────────────────────────────

const ROUNDS_TO_FETCH = 20;
const SCORES_CSV = path.resolve('../Men_s League 2026 Golfer Spreadsheet V2.csv');
const OUTPUT_JSON = path.resolve('./analysis-results.json');
const OUTPUT_CSV = path.resolve('./analysis-results.csv');

// Handicap flight thresholds — adjust for your field
const FLIGHT_THRESHOLDS = {
  A_MAX: 9.9,   // 0–9.9 = A flight
  B_MAX: 18.9,  // 10–18.9 = B flight
  // 19+ = C flight
};

// ── Types ─────────────────────────────────────────────────────────────────────

interface RosterPlayer {
  firstName: string;
  lastName: string;
  ghinId: string;
  handicapIndex: number;
  flight: 'A' | 'B' | 'C';
}

interface ScoreRecord {
  playedAt: string;
  adjustedGross: number;
  handicapDifferential: number;
  courseRating: number;
  slopeRating: number;
  holes: number;
}

interface PlayerAnalysis {
  name: string;
  ghinId: string;
  postedHandicap: number;
  flight: 'A' | 'B' | 'C';
  roundsFetched: number;
  avgDifferential: number;
  // Positive = playing BETTER than handicap (potential sandbagger)
  // Negative = playing WORSE than handicap
  diffVsHandicap: number;
  sandbaggerScore: number;   // 0–10, higher = more suspicious
  volatility: number;        // std dev of differentials — high = inconsistent
  trend: 'improving' | 'declining' | 'stable';
  recentDifferentials: number[];
  error?: string;
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function stdDev(values: number[]): number {
  if (values.length < 2) return 0;
  const mean = values.reduce((a, b) => a + b, 0) / values.length;
  const variance = values.map(v => Math.pow(v - mean, 2)).reduce((a, b) => a + b, 0) / values.length;
  return Math.sqrt(variance);
}

function mean(values: number[]): number {
  if (!values.length) return 0;
  return values.reduce((a, b) => a + b, 0) / values.length;
}

function trendDirection(differentials: number[]): 'improving' | 'declining' | 'stable' {
  if (differentials.length < 4) return 'stable';
  const firstHalf = mean(differentials.slice(0, Math.floor(differentials.length / 2)));
  const secondHalf = mean(differentials.slice(Math.floor(differentials.length / 2)));
  const delta = firstHalf - secondHalf; // more recent rounds come first in GHIN response
  if (delta > 1.5) return 'improving';   // recent rounds lower = improving
  if (delta < -1.5) return 'declining';
  return 'stable';
}

function sandbaggerScore(diffVsHandicap: number, volatility: number, roundsBelowHandicap: number, totalRounds: number): number {
  if (totalRounds === 0) return 0;
  // How far they play below their index (positive = beating handicap)
  const gapScore = Math.min(Math.max(diffVsHandicap * 1.5, 0), 5);
  // What % of rounds they beat their handicap by 2+ strokes
  const frequencyScore = (roundsBelowHandicap / totalRounds) * 3;
  // High volatility is slightly suspicious (sandbagging often shows volatility)
  const volatilityScore = Math.min(volatility / 5, 2);
  return Math.min(Math.round((gapScore + frequencyScore + volatilityScore) * 10) / 10, 10);
}

function assignFlight(handicap: number): 'A' | 'B' | 'C' {
  if (handicap <= FLIGHT_THRESHOLDS.A_MAX) return 'A';
  if (handicap <= FLIGHT_THRESHOLDS.B_MAX) return 'B';
  return 'C';
}

// ── Load roster ───────────────────────────────────────────────────────────────

function loadRoster(): RosterPlayer[] {
  const raw = fs.readFileSync(SCORES_CSV, 'utf8');
  const { data } = Papa.parse<Record<string, string>>(raw, { header: true, skipEmptyLines: true });

  return data
    .filter(row => row['GHIN Id'] && row['Index'] && row['GHIN Active'] === 'Yes')
    .map(row => {
      const handicap = parseFloat(row['Index']);
      return {
        firstName: row['First Name']?.trim() ?? '',
        lastName: row['Last Name']?.trim() ?? '',
        ghinId: row['GHIN Id'].trim(),
        handicapIndex: handicap,
        flight: assignFlight(handicap),
      };
    });
}

// ── Fetch & analyze one player ────────────────────────────────────────────────

async function analyzePlayer(
  client: GhinClient,
  player: RosterPlayer,
): Promise<PlayerAnalysis> {
  const base: Omit<PlayerAnalysis, 'roundsFetched' | 'avgDifferential' | 'diffVsHandicap' | 'sandbaggerScore' | 'volatility' | 'trend' | 'recentDifferentials'> = {
    name: `${player.firstName} ${player.lastName}`,
    ghinId: player.ghinId,
    postedHandicap: player.handicapIndex,
    flight: player.flight,
  };

  try {
    const scoresResponse = await client.golfers.getScores(player.ghinId, {
      limit: ROUNDS_TO_FETCH,
      offset: 0,
    });

    const scores: ScoreRecord[] = (scoresResponse ?? [])
      .filter((s: any) => s.handicap_differential != null && s.number_of_holes >= 18)
      .map((s: any) => ({
        playedAt: s.played_at,
        adjustedGross: s.adjusted_gross_score,
        handicapDifferential: parseFloat(s.handicap_differential),
        courseRating: parseFloat(s.course_rating),
        slopeRating: parseFloat(s.slope_rating),
        holes: s.number_of_holes,
      }));

    if (!scores.length) {
      return { ...base, roundsFetched: 0, avgDifferential: player.handicapIndex, diffVsHandicap: 0, sandbaggerScore: 0, volatility: 0, trend: 'stable', recentDifferentials: [] };
    }

    const differentials = scores.map(s => s.handicapDifferential);
    const avg = mean(differentials);
    const vol = stdDev(differentials);

    // diffVsHandicap: positive means they're playing BETTER than their posted index
    const diffVsHandicap = parseFloat((player.handicapIndex - avg).toFixed(2));

    // Rounds where differential was 2+ strokes better than handicap
    const roundsBelowHandicap = differentials.filter(d => d <= player.handicapIndex - 2).length;

    return {
      ...base,
      roundsFetched: scores.length,
      avgDifferential: parseFloat(avg.toFixed(2)),
      diffVsHandicap,
      sandbaggerScore: sandbaggerScore(diffVsHandicap, vol, roundsBelowHandicap, scores.length),
      volatility: parseFloat(vol.toFixed(2)),
      trend: trendDirection(differentials),
      recentDifferentials: differentials.slice(0, 10),
    };
  } catch (err: any) {
    return {
      ...base,
      roundsFetched: 0,
      avgDifferential: player.handicapIndex,
      diffVsHandicap: 0,
      sandbaggerScore: 0,
      volatility: 0,
      trend: 'stable',
      recentDifferentials: [],
      error: err?.message ?? 'unknown error',
    };
  }
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  if (!process.env.GHIN_USERNAME || !process.env.GHIN_PASSWORD) {
    console.error('❌  Set GHIN_USERNAME and GHIN_PASSWORD in .env');
    process.exit(1);
  }

  const client = new GhinClient({
    username: process.env.GHIN_USERNAME,
    password: process.env.GHIN_PASSWORD,
  });

  const roster = loadRoster();
  console.log(`\n📋  Loaded ${roster.length} active players from roster`);
  console.log(`    A flight (0–9.9):   ${roster.filter(p => p.flight === 'A').length} players`);
  console.log(`    B flight (10–18.9): ${roster.filter(p => p.flight === 'B').length} players`);
  console.log(`    C flight (19+):     ${roster.filter(p => p.flight === 'C').length} players\n`);

  const results: PlayerAnalysis[] = [];
  let done = 0;

  for (const player of roster) {
    process.stdout.write(`\r  Fetching ${++done}/${roster.length}: ${player.firstName} ${player.lastName}`.padEnd(60));
    const result = await analyzePlayer(client, player);
    results.push(result);
    await new Promise(r => setTimeout(r, 150)); // be polite to the API
  }

  console.log('\n\n✅  Done fetching. Writing results...\n');

  // Sort by sandbagger score desc within each flight
  results.sort((a, b) => {
    if (a.flight !== b.flight) return a.flight.localeCompare(b.flight);
    return b.sandbaggerScore - a.sandbaggerScore;
  });

  // Write JSON
  fs.writeFileSync(OUTPUT_JSON, JSON.stringify(results, null, 2));

  // Write CSV
  const csvRows = results.map(r => ({
    Flight: r.flight,
    Name: r.name,
    'GHIN ID': r.ghinId,
    'Posted HCP': r.postedHandicap,
    'Avg Differential': r.avgDifferential,
    'Diff vs HCP (+= beats HCP)': r.diffVsHandicap,
    'Sandbagger Score (0-10)': r.sandbaggerScore,
    Volatility: r.volatility,
    Trend: r.trend,
    'Rounds Fetched': r.roundsFetched,
    'Recent 10 Differentials': r.recentDifferentials.join(', '),
    Error: r.error ?? '',
  }));
  fs.writeFileSync(OUTPUT_CSV, Papa.unparse(csvRows));

  // Print top suspects per flight
  for (const flight of ['A', 'B', 'C'] as const) {
    const flightPlayers = results.filter(r => r.flight === flight && r.roundsFetched >= 5);
    if (!flightPlayers.length) continue;

    console.log(`═══ Flight ${flight} — Top Sandbagger Suspects ═══`);
    flightPlayers.slice(0, 5).forEach((p, i) => {
      const indicator = p.sandbaggerScore >= 7 ? '🚨' : p.sandbaggerScore >= 4 ? '⚠️ ' : '✅';
      console.log(
        `  ${i + 1}. ${indicator} ${p.name.padEnd(25)} HCP: ${String(p.postedHandicap).padEnd(5)} Avg Diff: ${String(p.avgDifferential).padEnd(6)} Delta: ${p.diffVsHandicap > 0 ? '+' : ''}${p.diffVsHandicap} | Score: ${p.sandbaggerScore}/10 | ${p.trend}`
      );
    });
    console.log('');
  }

  console.log(`📁  Full results saved to:`);
  console.log(`    ${OUTPUT_JSON}`);
  console.log(`    ${OUTPUT_CSV}\n`);
}

// Load .env manually (no dotenv dep needed)
const envPath = new URL('.env', import.meta.url).pathname;
if (fs.existsSync(envPath)) {
  fs.readFileSync(envPath, 'utf8').split('\n').forEach(line => {
    const [k, ...v] = line.split('=');
    if (k && v.length) process.env[k.trim()] = v.join('=').trim();
  });
}

main().catch(err => {
  console.error('\n❌  Fatal error:', err.message);
  process.exit(1);
});

File: analyze.py
#!/usr/bin/env python3
"""
Calcutta Analysis Tool — Youche CC

Pulls last 20 rounds for each player from GHIN, computes:
  - Average handicap differential vs posted handicap index
  - Sandbagger score (0-10)
  - Volatility (std dev of differentials)
  - Trend (improving / declining / stable)
  - Flight assignment (A / B / C)

Usage:
  cp .env.example .env    # fill in GHIN_USERNAME and GHIN_PASSWORD
  python3 analyze.py
"""

import csv
import json
import os
import sys
import time
import statistics
import urllib.request
import urllib.parse
import urllib.error
from pathlib import Path
from typing import Optional

# ── Config ────────────────────────────────────────────────────────────────────

ROUNDS_TO_FETCH = 20
ROSTER_CSV = Path(__file__).parent.parent / "Men_s League 2026 Golfer Spreadsheet V2.csv"
OUTPUT_JSON = Path(__file__).parent / "analysis-results.json"
OUTPUT_CSV  = Path(__file__).parent / "analysis-results.csv"

FLIGHT_A_MAX = 9.9
FLIGHT_B_MAX = 18.9

GHIN_API_BASE = "https://api2.ghin.com/api/v1"

# ── Load .env ─────────────────────────────────────────────────────────────────

def load_env():
    env_path = Path(__file__).parent / ".env"
    if env_path.exists():
        for line in env_path.read_text().splitlines():
            line = line.strip()
            if line and not line.startswith("#") and "=" in line:
                k, _, v = line.partition("=")
                os.environ.setdefault(k.strip(), v.strip())

# ── GHIN API client ───────────────────────────────────────────────────────────

FIREBASE_URL = "https://firebaseinstallations.googleapis.com/v1/projects/ghin-mobile-app/installations"
FIREBASE_API_KEY = "AIzaSyBxgTOAWxiud0HuaE5tN-5NTlzFnrtyz-I"
FIREBASE_PAYLOAD = {
    "appId": "1:884417644529:web:47fb315bc6c70242f72650",
    "authVersion": "FIS_v2",
    "sdkVersion": "w:0.5.7",
    "fid": "fg6JfS0U01YmrelthLX9Iz",
}

class GhinClient:
    def __init__(self, username: str, password: str):
        self.token: Optional[str] = None
        self._login(username, password)

    def _http(self, method: str, url: str, body: Optional[dict] = None,
              params: Optional[dict] = None, headers: Optional[dict] = None) -> dict:
        if params:
            url += "?" + urllib.parse.urlencode(params)
        data = json.dumps(body).encode() if body else None
        hdrs = {"Content-Type": "application/json", "Accept": "application/json"}
        if headers:
            hdrs.update(headers)
        if self.token:
            hdrs["Authorization"] = f"Bearer {self.token}"
        req = urllib.request.Request(url, data=data, headers=hdrs, method=method)
        try:
            with urllib.request.urlopen(req, timeout=15) as resp:
                return json.loads(resp.read().decode())
        except urllib.error.HTTPError as e:
            raise RuntimeError(f"HTTP {e.code} {url}: {e.read().decode()[:200]}")

    def _login(self, username: str, password: str):
        # Step 1: get Firebase session token
        fb = self._http("POST", FIREBASE_URL, body=FIREBASE_PAYLOAD,
                        headers={"x-goog-api-key": FIREBASE_API_KEY})
        firebase_token = fb["authToken"]["token"]

        # Step 2: GHIN login
        resp = self._http("POST", f"{GHIN_API_BASE}/golfer_login.json", body={
            "token": firebase_token,
            "user": {"email_or_ghin": username, "password": password},
        })
        self.token = resp.get("golfer_user", {}).get("golfer_user_token")
        if not self.token:
            raise RuntimeError(f"Login failed — no token in response: {list(resp.keys())}")

    def get_scores(self, ghin_number: str, limit: int = 20) -> list[dict]:
        resp = self._http("GET", f"{GHIN_API_BASE}/golfers/{ghin_number}/scores.json",
                          params={"limit": limit, "offset": 0})
        # API returns scores split across recent_scores and revision_scores dicts
        recent   = resp.get("recent_scores",   {}).get("scores", []) if isinstance(resp.get("recent_scores"),   dict) else []
        revision = resp.get("revision_scores", {}).get("scores", []) if isinstance(resp.get("revision_scores"), dict) else []
        return recent + revision

    def get_handicap(self, ghin_number: str) -> Optional[float]:
        try:
            resp = self._http("GET", f"{GHIN_API_BASE}/golfers/search.json",
                              params={"golfer_id": ghin_number, "page": 1, "per_page": 5})
            golfers = resp.get("golfers", [])
            if not golfers:
                return None
            return float(golfers[0]["hi_value"])
        except Exception:
            return None

# ── Helpers ───────────────────────────────────────────────────────────────────

def assign_flight(hcp: float) -> str:
    if hcp <= FLIGHT_A_MAX: return "A"
    if hcp <= FLIGHT_B_MAX: return "B"
    return "C"

def calc_trend(diffs: list[float]) -> str:
    if len(diffs) < 4:
        return "stable"
    mid = len(diffs) // 2
    # GHIN returns most recent first
    recent_avg = statistics.mean(diffs[:mid])
    older_avg  = statistics.mean(diffs[mid:])
    delta = older_avg - recent_avg  # positive = recent scores lower = improving
    if delta > 1.5: return "improving"
    if delta < -1.5: return "declining"
    return "stable"

def sandbagger_score(diff_vs_hcp: float, vol: float, rounds_below: int, total: int) -> float:
    if total == 0: return 0.0
    gap_score  = min(max(diff_vs_hcp * 1.5, 0), 5)
    freq_score = (rounds_below / total) * 3
    vol_score  = min(vol / 5, 2)
    return round(min(gap_score + freq_score + vol_score, 10), 1)

def flag_label(s: float, error: bool) -> str:
    if error:        return "❓ NO DATA"
    if s >= 7:       return "🚨 SANDBAG "
    if s >= 4:       return "⚠️  WATCH   "
    return           "✅ CLEAN   "

# ── Load roster ───────────────────────────────────────────────────────────────

def load_roster() -> list[dict]:
    players = []
    with open(ROSTER_CSV, newline="", encoding="utf-8-sig") as f:
        for row in csv.DictReader(f):
            ghin = row.get("GHIN Id", "").strip()
            idx  = row.get("Index", "").strip()
            active = row.get("GHIN Active", "").strip()
            if not ghin or not idx or active != "Yes":
                continue
            try:
                hcp = float(idx)
            except ValueError:
                continue
            players.append({
                "name":     f"{row.get('First Name','').strip()} {row.get('Last Name','').strip()}".strip(),
                "ghin_id":  ghin,
                "posted_hcp": hcp,
                "flight":   assign_flight(hcp),
            })
    return players

# ── Analyze one player ────────────────────────────────────────────────────────

def analyze_player(client: GhinClient, player: dict) -> dict:
    result = {
        "name":           player["name"],
        "ghin_id":        player["ghin_id"],
        "posted_hcp":     player["posted_hcp"],
        "flight":         player["flight"],
        "rounds_fetched": 0,
        "avg_differential": player["posted_hcp"],
        "diff_vs_hcp":    0.0,
        "sandbagger_score": 0.0,
        "volatility":     0.0,
        "trend":          "stable",
        "recent_diffs":   [],
        "error":          None,
    }

    try:
        scores = client.get_scores(player["ghin_id"], limit=ROUNDS_TO_FETCH)
        valid = [
            float(s["differential"])
            for s in scores
            if s.get("differential") is not None
               and s.get("number_of_holes", 18) >= 18
               and s.get("status", "").lower() == "validated"
        ]

        if not valid:
            result["error"] = "no 18-hole scores"
            return result

        avg = statistics.mean(valid)
        vol = statistics.stdev(valid) if len(valid) > 1 else 0.0
        diff_vs_hcp = round(player["posted_hcp"] - avg, 2)
        rounds_below = sum(1 for d in valid if d <= player["posted_hcp"] - 2)

        result.update({
            "rounds_fetched":   len(valid),
            "avg_differential": round(avg, 2),
            "diff_vs_hcp":      diff_vs_hcp,
            "sandbagger_score": sandbagger_score(diff_vs_hcp, vol, rounds_below, len(valid)),
            "volatility":       round(vol, 2),
            "trend":            calc_trend(valid),
            "recent_diffs":     valid[:10],
        })
    except Exception as e:
        result["error"] = str(e)

    return result

# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    load_env()

    username = os.environ.get("GHIN_USERNAME")
    password = os.environ.get("GHIN_PASSWORD")

    if not username or not password:
        print("❌  Create a .env file with GHIN_USERNAME and GHIN_PASSWORD")
        print("    See .env.example for the format")
        sys.exit(1)

    print("\n🔐  Logging in to GHIN...")
    try:
        client = GhinClient(username, password)
        print("✅  Authenticated\n")
    except RuntimeError as e:
        print(f"❌  {e}")
        sys.exit(1)

    roster = load_roster()
    flights = {f: sum(1 for p in roster if p["flight"] == f) for f in "ABC"}
    print(f"📋  Loaded {len(roster)} active players from roster")
    print(f"    A flight (0–{FLIGHT_A_MAX}):   {flights['A']} players")
    print(f"    B flight ({FLIGHT_A_MAX+0.1:.1f}–{FLIGHT_B_MAX}): {flights['B']} players")
    print(f"    C flight ({FLIGHT_B_MAX+0.1:.1f}+):     {flights['C']} players\n")

    results = []
    for i, player in enumerate(roster, 1):
        print(f"\r  Fetching {i}/{len(roster)}: {player['name']:<30}", end="", flush=True)
        results.append(analyze_player(client, player))
        time.sleep(0.15)  # be polite to the API

    print("\n\n✅  Done. Writing results...\n")

    # Sort: by flight, then sandbagger score desc
    results.sort(key=lambda r: (r["flight"], -r["sandbagger_score"]))

    OUTPUT_JSON.write_text(json.dumps(results, indent=2))

    with open(OUTPUT_CSV, "w", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=[
            "flight", "name", "ghin_id", "posted_hcp", "avg_differential",
            "diff_vs_hcp", "sandbagger_score", "volatility", "trend",
            "rounds_fetched", "recent_diffs", "error"
        ])
        writer.writeheader()
        for r in results:
            writer.writerow({**r, "recent_diffs": ", ".join(str(d) for d in r["recent_diffs"])})

    # Print top suspects
    for flight in "ABC":
        flight_players = [r for r in results if r["flight"] == flight and r["rounds_fetched"] >= 5]
        if not flight_players:
            continue
        print(f"{'═'*70}")
        print(f"  Flight {flight} — Top Sandbagger Suspects")
        print(f"{'═'*70}")
        for i, p in enumerate(flight_players[:5], 1):
            delta = f"+{p['diff_vs_hcp']}" if p["diff_vs_hcp"] > 0 else str(p["diff_vs_hcp"])
            print(f"  {i}. {flag_label(p['sandbagger_score'], bool(p['error']))} "
                  f"{p['name']:<24} HCP:{p['posted_hcp']:<6} "
                  f"AvgDiff:{p['avg_differential']:<7} Δ:{delta:<7} "
                  f"Score:{p['sandbagger_score']}/10 | {p['trend']}")
        print()

    errors = [r for r in results if r["error"]]
    if errors:
        print(f"⚠️   {len(errors)} players had errors (no scores / API issue):")
        for r in errors[:5]:
            print(f"    - {r['name']} ({r['ghin_id']}): {r['error']}")
        print()

    print(f"📁  Results saved to:")
    print(f"    {OUTPUT_JSON}")
    print(f"    {OUTPUT_CSV}\n")

if __name__ == "__main__":
    main()
